File: sdks/java/src/main/java/io/coordin8/DjinnClient.java
package io.coordin8;

import io.grpc.ManagedChannel;
import io.grpc.ManagedChannelBuilder;

import java.util.Map;
import java.util.Optional;
import java.util.concurrent.TimeUnit;

/**
 * Entry point for all Djinn interactions.
 *
 * <pre>{@code
 * try (DjinnClient djinn = DjinnClient.connect("localhost:9002")) {
 *     LeaseClient    leases   = djinn.registryLeases();
 *     RegistryClient registry = djinn.registry();
 *     ProxyClient    proxy    = djinn.proxy();
 * }
 * }</pre>
 *
 * <p>There is no single "LeaseMgr" to connect to — leasing is distributed.
 * Registry, Space, and EventMgr each grant their own leases and mount
 * {@code LeaseService} on their own connection; renew/cancel a lease by
 * dialing whichever one granted it (its address is carried on the lease
 * itself, in {@code grantorHost}/{@code grantorPort}) via
 * {@link LeaseClient#dial(String)}. See
 * {@code .claude/plans/distributed-leasing/PRD.md}.
 */
public class DjinnClient implements AutoCloseable {

    private final ManagedChannel registryChannel;
    private final ManagedChannel proxyChannel;
    private final ManagedChannel spaceChannel;
    private final ManagedChannel eventChannel;

    private DjinnClient(ManagedChannel registryChannel,
                        ManagedChannel proxyChannel,
                        ManagedChannel spaceChannel,
                        ManagedChannel eventChannel) {
        this.registryChannel = registryChannel;
        this.proxyChannel    = proxyChannel;
        this.spaceChannel    = spaceChannel;
        this.eventChannel    = eventChannel;
    }

    /**
     * Dial Registry directly at {@code registryAddr} — the one address a
     * caller needs to know in advance — then look up Proxy, Space, and
     * EventMgr through it, the same way any application service is
     * discovered via {@link ServiceDiscovery}. Works identically against a
     * bundled monolith or a fully split, multi-host deployment: Registry
     * just returns whatever address each service actually registered.
     *
     * @param registryAddr Registry's address, {@code "host:port"}
     */
    public static DjinnClient connect(String registryAddr) {
        return connect(registryAddr, null, null, null);
    }

    /**
     * Same as {@link #connect(String)}, but lets a caller pin Proxy/Space/
     * EventMgr's address explicitly instead of looking it up through
     * Registry — an escape hatch mirroring the Go SDK's
     * {@code WithProxyAddr}/{@code WithSpaceAddr}/{@code WithEventAddr}.
     * Pass {@code null} for any address that should be resolved via
     * {@code Registry.Lookup} instead.
     *
     * @param registryAddr Registry's address, {@code "host:port"}
     * @param proxyAddr    pinned Proxy address, or {@code null} to look up
     * @param spaceAddr    pinned Space address, or {@code null} to look up
     * @param eventAddr    pinned EventMgr address, or {@code null} to look up
     */
    public static DjinnClient connect(String registryAddr, String proxyAddr,
                                      String spaceAddr, String eventAddr) {
        ManagedChannel registryChannel = channelFor(registryAddr);
        ManagedChannel proxyChannel = null;
        ManagedChannel spaceChannel = null;
        ManagedChannel eventChannel = null;
        try {
            RegistryClient registryClient = new RegistryClient(registryChannel);

            proxyChannel = channelFor(resolve(registryClient, proxyAddr, "Proxy"));
            spaceChannel = channelFor(resolve(registryClient, spaceAddr, "Space"));
            eventChannel = channelFor(resolve(registryClient, eventAddr, "EventMgr"));

            return new DjinnClient(registryChannel, proxyChannel, spaceChannel, eventChannel);
        } catch (RuntimeException e) {
            registryChannel.shutdown();
            if (proxyChannel != null) proxyChannel.shutdown();
            if (spaceChannel != null) spaceChannel.shutdown();
            if (eventChannel != null) eventChannel.shutdown();
            throw e;
        }
    }

    private static String resolve(RegistryClient registryClient, String pinned, String interfaceName) {
        if (pinned != null) {
            return pinned;
        }
        Optional<RegistryClient.CapabilityRecord> cap =
                registryClient.lookup(Map.of("interface", interfaceName));
        RegistryClient.CapabilityRecord record = cap.orElseThrow(() ->
                new IllegalStateException("look up " + interfaceName + ": not found in registry"));
        RegistryClient.TransportDescriptor transport = record.transport();
        if (transport == null) {
            throw new IllegalStateException("look up " + interfaceName + ": no transport in registry entry");
        }
        String host = transport.config().get("host");
        String port = transport.config().get("port");
        if (host == null || host.isEmpty() || port == null || port.isEmpty()) {
            throw new IllegalStateException("look up " + interfaceName + ": missing host/port in transport config");
        }
        return host + ":" + port;
    }

    private static ManagedChannel channelFor(String addr) {
        int idx = addr.lastIndexOf(':');
        if (idx < 0) {
            throw new IllegalArgumentException("address must be host:port, got: " + addr);
        }
        String host = addr.substring(0, idx);
        int port = Integer.parseInt(addr.substring(idx + 1));
        return ManagedChannelBuilder.forAddress(host, port).usePlaintext().build();
    }

    /**
     * Returns a {@link LeaseClient} for renewing/cancelling leases that
     * Registry itself granted — every self-registration via
     * {@code registry().register(...)} returns one. {@code LeaseService} is
     * mounted on Registry's own connection (no separate dial needed),
     * matching how Registry embeds its own {@code LeaseManager}.
     */
    public LeaseClient registryLeases() { return LeaseClient.wrap(registryChannel); }

    public RegistryClient registry() { return new RegistryClient(registryChannel); }
    public ProxyClient    proxy()    { return new ProxyClient(proxyChannel); }
    public SpaceClient    space()    { return new SpaceClient(spaceChannel); }
    public EventClient    events()   { return new EventClient(eventChannel); }

    @Override
    public void close() throws InterruptedException {
        registryChannel.shutdown().awaitTermination(5, TimeUnit.SECONDS);
        proxyChannel.shutdown().awaitTermination(5, TimeUnit.SECONDS);
        spaceChannel.shutdown().awaitTermination(5, TimeUnit.SECONDS);
        eventChannel.shutdown().awaitTermination(5, TimeUnit.SECONDS);
    }
}

File: sdks/node/src/djinn-client.ts
import * as grpc from "@grpc/grpc-js";
import { LeaseClient } from "./lease-client";
import { RegistryClient } from "./registry-client";
import { ProxyClient } from "./proxy-client";
import { SpaceClient } from "./space-client";
import { EventClient } from "./event-client";
import { interceptorsFor } from "./auth";

/**
 * Options for DjinnClient.connect(). Use these to pin a specific service's
 * address instead of looking it up through Registry.
 */
export interface ConnectOptions {
  /** Pin Proxy's address instead of looking it up through Registry. */
  proxyAddr?: string;
  /** Pin Space's address instead of looking it up through Registry. */
  spaceAddr?: string;
  /** Pin EventMgr's address instead of looking it up through Registry. */
  eventAddr?: string;
  /**
   * Bearer token attached to every call on every connection — required
   * against a Djinn with COORDIN8_JWT_SECRET set, harmless (ignored)
   * against one that doesn't have auth enabled. Mint one with
   * `coordin8 auth mint-token`; see .claude/plans/grpc-security/PRD.md.
   */
  token?: string;
}

/**
 * DjinnClient is the entry point for all Djinn interactions.
 *
 * There is no single "LeaseMgr" to connect to — leasing is distributed.
 * Registry, Space, and EventMgr each grant their own leases and mount
 * LeaseService on their own connection; renew/cancel a lease by dialing
 * whichever one granted it (its address is carried on the Lease itself, in
 * `grantorHost`/`grantorPort`) via `dialLease()`. See
 * .claude/plans/distributed-leasing/PRD.md.
 */
export class DjinnClient {
  private readonly registryChannel: grpc.Channel;
  private readonly proxyChannel: grpc.Channel;
  private readonly spaceChannel: grpc.Channel;
  private readonly eventChannel: grpc.Channel;
  private readonly interceptors: grpc.Interceptor[];
  private readonly _registry: RegistryClient;
  private readonly _proxy: ProxyClient;
  private readonly _space: SpaceClient;
  private readonly _event: EventClient;

  private constructor(
    registryChannel: grpc.Channel,
    proxyChannel: grpc.Channel,
    spaceChannel: grpc.Channel,
    eventChannel: grpc.Channel,
    interceptors: grpc.Interceptor[]
  ) {
    this.registryChannel = registryChannel;
    this.proxyChannel = proxyChannel;
    this.spaceChannel = spaceChannel;
    this.eventChannel = eventChannel;
    this.interceptors = interceptors;
    this._registry = new RegistryClient(registryChannel, interceptors);
    this._proxy = new ProxyClient(proxyChannel, interceptors);
    this._space = new SpaceClient(spaceChannel, interceptors);
    this._event = new EventClient(eventChannel, interceptors);
  }

  /**
   * Dials Registry directly at registryAddr — the one address a caller
   * needs to know in advance — then looks up Proxy, Space, and EventMgr
   * through it, the same way any application service is discovered via
   * ServiceDiscovery. Works identically against a bundled monolith or a
   * fully split, multi-host deployment: Registry just returns whatever
   * address each service actually registered.
   *
   * Use opts.proxyAddr / opts.spaceAddr / opts.eventAddr to pin a specific
   * service's address instead of looking it up.
   */
  static async connect(registryAddr: string, opts: ConnectOptions = {}): Promise<DjinnClient> {
    const creds = grpc.credentials.createInsecure();
    const interceptors = interceptorsFor(opts.token);
    const registryChannel = new grpc.Channel(registryAddr, creds, {});
    const registry = new RegistryClient(registryChannel, interceptors);

    let proxyAddr: string;
    let spaceAddr: string;
    let eventAddr: string;
    try {
      [proxyAddr, spaceAddr, eventAddr] = await Promise.all([
        resolveAddr(registry, opts.proxyAddr, "Proxy"),
        resolveAddr(registry, opts.spaceAddr, "Space"),
        resolveAddr(registry, opts.eventAddr, "EventMgr"),
      ]);
    } catch (err) {
      registryChannel.close();
      throw err;
    }

    const proxyChannel = new grpc.Channel(proxyAddr, creds, {});
    const spaceChannel = new grpc.Channel(spaceAddr, creds, {});
    const eventChannel = new grpc.Channel(eventAddr, creds, {});

    return new DjinnClient(registryChannel, proxyChannel, spaceChannel, eventChannel, interceptors);
  }

  registry(): RegistryClient { return this._registry; }
  proxy():    ProxyClient    { return this._proxy; }
  space():    SpaceClient    { return this._space; }
  events():   EventClient    { return this._event; }

  /**
   * Returns a LeaseClient for renewing/cancelling leases that Registry
   * itself granted — every register() call returns one. LeaseService is
   * mounted on Registry's own connection (no separate dial needed),
   * matching how Registry embeds its own LeaseManager.
   */
  registryLeases(): LeaseClient {
    return LeaseClient.fromChannel(this.registryChannel, this.interceptors);
  }

  close(): void {
    this.registryChannel.close();
    this.proxyChannel.close();
    this.spaceChannel.close();
    this.eventChannel.close();
  }
}

/**
 * Looks up interfaceName in Registry and returns its "host:port" transport
 * address, unless pinned overrides the lookup.
 */
async function resolveAddr(
  registry: RegistryClient,
  pinned: string | undefined,
  interfaceName: string
): Promise<string> {
  if (pinned) return pinned;

  const record = await registry.lookup({ interface: interfaceName });
  if (!record) {
    throw new Error(`look up ${interfaceName}: not found in registry`);
  }
  const host = record.transport?.config?.host;
  const port = record.transport?.config?.port;
  if (!host || !port) {
    throw new Error(`look up ${interfaceName}: missing host/port in transport config`);
  }
  return `${host}:${port}`;
}
